runner/src/mill/runner/CodeGen.scala

```diff
@@ -0,0 +1,297 @@
+           |  $childAliases
+           |}""".stripMargin
+      case None =>
+        s"""$pkgLine
+           |$aliasImports
+           |$prelude
+           |${topBuildHeader(segments, scriptFolderPath, millTopLevelProjectRoot)}
+           |$markerComment
+           |$scriptCode
+           |// define this after user code so in case of conflict these lines are what turn
+           |// up in the error message, so we can add a comment and control what the user sees
+           |$childAliases
+           |}""".stripMargin
+
+    }
+  }
+
+  def topBuildPrelude(
+      scriptFolderPath: os.Path,
+      enclosingClasspath: Seq[os.Path],
+      millTopLevelProjectRoot: os.Path,
+      childAliases: String
+  ): String = {
+    s"""import _root_.mill.runner.MillBuildRootModule
+       |@_root_.scala.annotation.nowarn
+       |object MillMiscInfo extends MillBuildRootModule.MillMiscInfo(
+       |  ${enclosingClasspath.map(p => literalize(p.toString))},
+       |  ${literalize(scriptFolderPath.toString)},
+       |  ${literalize(millTopLevelProjectRoot.toString)},
+       |  _root_.mill.define.Discover[$wrapperObjectName.type]
+       |){
+       |  // aliases so child modules can be referred to directly as `foo` rather
+       |  // than `foo.module`. Need to be outside `MillPackageClass` in case they are
+       |  // referenced in the combined `extends` clause
+       |  $childAliases
+       |}
+       |import MillMiscInfo._
+       |""".stripMargin
+  }
+
+  def topBuildHeader(
+      segs: Seq[String],
+      scriptFolderPath: os.Path,
+      millTopLevelProjectRoot: os.Path
+  ): String = {
+    val extendsClause = if (segs.isEmpty) {
+      if (millTopLevelProjectRoot == scriptFolderPath) {
+        s"extends _root_.mill.main.RootModule() "
+      } else {
+        s"extends _root_.mill.runner.MillBuildRootModule() "
+      }
+    } else {
+      val segsList = segs.map(pprint.Util.literalize(_)).mkString(", ")
+      s"extends _root_.mill.main.RootModule.Subfolder($segsList) "
+    }
+
+    // User code needs to be put in a separate class for proper submodule
+    // object initialization due to https://github.com/scala/scala3/issues/21444
+    s"""object $wrapperObjectName extends $wrapperObjectName
+       |class $wrapperObjectName $extendsClause {""".stripMargin
+
+  }
+
+  private case class Snippet(var text: String = null, var start: Int = -1, var end: Int = -1) {
+    def applyTo(s: String, replacement: String): String =
+      s.patch(start, replacement.padTo(end - start, ' '), end - start)
+  }
+
+  private case class ObjectData(obj: Snippet, name: Snippet, parent: Snippet)
+
+  // Use Fastparse's Instrument API to identify top-level `object`s during a parse
+  // and fish out the start/end indices and text for parts of the code that we need
+  // to mangle and replace
+  private class ObjectDataInstrument(scriptCode: String) extends fastparse.internal.Instrument {
```

Comment:
CC @bishabosha this parsing logic may need to be ported to the Scala 3 port. Not sure how the Scala 3 `import $ivy` parser works, but I assume you have some kind of AST, which may make things a lot easier than the visitor-style parsing I have to do here using ScalaParse

Reply:
Thanks, this is probably next for me